### PR TITLE
feat: Unified caching interface with consistent TTL and refresh (Issue #154)

### LINKED ISSUE #154: refactor: Unified caching interface for URL and response caches
## Overview

URL list caches vary (URLCache, RichURLCache); only Walmart caches responses. Refresh semantics differ (`--refresh-urls` only on some scrapers).

## Proposal

Create unified caching interface:
- Abstract cache interface with TTL + `--refresh` flag
- Plug in: Walmart's HTML cache, AT&T/Target URL caches
- Optional response caching for Telus/BestBuy
- Consistent refresh behavior across scrapers

## Priority
⚪ **Tier 4: Optional/Later** - Nice to have after runner/serializer

## Depends On
- Shared scrape runner
- Clear understanding of each scraper's caching needs

## Acceptance Criteria
- [ ] Unified cache interface defined
- [ ] Existing caches migrated to interface
- [ ] `--refresh` flag behavior consistent
- [ ] TTL configurable per cache type

### PR DESCRIPTION
**Pull request by vmud**
## Summary

Implements Issue #154 by creating a unified caching interface that provides consistent TTL and refresh semantics across all cache types.

### Changes

- **New `src/shared/cache_interface.py`**: Abstract `CacheInterface` base class with generic typing
- **URLListCache**: Cache for simple URL lists (sitemap URLs, store URLs)
- **RichURLCache**: Cache for URLs with metadata (store_id, slug, etc.)
- **ResponseCache**: Cache for HTTP responses (HTML, JSON) - generalizes Walmart-specific caching
- **Consistent API**: All caches use `get(identifier, force_refresh)` and `set(identifier, data)`
- **Unified TTL**: Configurable TTL with automatic expiry checking
- **Metadata tracking**: Cache age, expiry status, timestamps
- **Migration guide**: Complete documentation in `docs/cache-interface-migration.md`

### Test Coverage

- 33 new unit tests covering all cache types and interface methods
- 100% pass rate
- Tests include expiry, force refresh, serialization, metadata, and integration workflows

### Backward Compatibility

The legacy `URLCache` and `RichURLCache` classes in `src/shared/cache.py` remain unchanged for backward compatibility. Scrapers can migrate incrementally.

### Benefits

1. **Consistent `--refresh-urls` behavior**: All scrapers can use the same flag semantics
2. **Type safety**: Generic typing provides better IDE support and type checking
3. **Extensible**: Easy to add new cache types (StateCache, APIResponseCache, etc.)
4. **Testable**: Abstract interface makes testing easier
5. **Maintainable**: Centralized cache logic reduces duplication

### Migration Path

See `docs/cache-interface-migration.md` for complete migration examples. Key changes:

**Before:**
```python
from src.shared.cache import URLCache
cache = URLCache('target')
urls = cache.get()
```

**After:**
```python
from src.shared.cache_interface import URLListCache
cache = URLListCache('target')
urls = cache.get('target', force_refresh=kwargs.get('refresh_urls', False))
```

### Future Work

This interface enables:
- Consistent `--refresh` flag for response caching across all scrapers
- Easy addition of new cache types (state data, API responses, sitemaps)
- Centralized cache management utilities
- Cache warming and preloading strategies

Fixes #154

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Introduces new cache-file format/paths and new public exports; while legacy caches remain, incorrect adoption could change caching/refresh behavior and lead to extra requests or stale data if scrapers migrate incorrectly.
> 
> **Overview**
> Adds a new unified caching layer (`CacheInterface`) with consistent `get(identifier, force_refresh)`/`set(identifier, data)` semantics, TTL-based expiry, and metadata (`cached_at`, age, expired) shared across cache types.
> 
> Introduces concrete caches for URL lists (`URLListCache`), URL+metadata lists (`RichURLCache`), and raw HTTP responses (`ResponseCache`, generalizing prior scraper-specific response caching), while keeping legacy `URLCache`/`RichURLCache` intact and re-exporting the new API via `src/shared/__init__.py`. Includes a migration guide (`docs/cache-interface-migration.md`) and a comprehensive new unit test suite (`tests/test_cache_interface.py`) covering expiry edge cases, corruption handling, and force-refresh behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5534fde148bdba25ed229426d7068de9ca3f4b18. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/shared/cache_interface.py [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+            # Extract the actual data (excluding metadata)
+            data_str = cache_data.get('data')
+            if data_str is None:
+                return None
+
+            return self.deserialize(data_str)
+
+        except (json.JSONDecodeError, KeyError, ValueError) as e:
+            logging.warning(f"Error reading cache for {identifier}: {e}")
+            return None
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/cache_interface.py]: Fixed in commit 48529f8. Refactored `get()` to use internal `cached_at` timestamp instead of unreliable file mtime. Also updated `is_valid()` for consistency. The `_is_expired()` method was removed since it's no longer used.

File: src/shared/cache_interface.py [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    def get_cache_key(self, url: str) -> str:
+        """Generate cache key from URL hash.
+
+        Args:
+            url: Full URL to cache
+
+        Returns:
+            SHA256 hash prefix (16 chars)
+        """
+        return hashlib.sha256(url.encode()).hexdigest()[:16]
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/cache_interface.py]: Fixed in commit 48529f8. Now using full 64-char SHA256 hash instead of truncated 16-char prefix. While the collision risk was low at current scale, using the full hash eliminates any potential for collisions regardless of cache size.

File: src/shared/cache_interface.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+            if not cached_at:
+                return None
+
+            cache_time = datetime.fromisoformat(cached_at)
+            age_days = (datetime.now() - cache_time).days
+
+            return {
+                'cached_at': cached_at,
+                'age_days': age_days,
+                'expired': age_days * 86400 > self.ttl.total_seconds()
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/cache_interface.py]: Fixed in commit 48529f8. The `get_metadata()` method now uses `age > self.ttl` (timedelta comparison) which matches the precision of `get()` and `is_valid()`. Added edge case tests to verify consistency between all three methods.

File: src/shared/cache_interface.py [hunk 4]
```diff
@@ -0,0 +1,438 @@
+"""Unified caching interface for URL and response caches - Issue #154.
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/cache_interface.py]: Fixed in commit 48529f8. Added `cache_interface.py` to the Architecture section of CLAUDE.md with description: 'Unified caching interface with consistent TTL (#154)'.

File: tests/test_cache_interface.py [hunk 5]
```diff
@@ -0,0 +1,452 @@
+"""Unit tests for unified cache interface - Issue #154."""
```
[the remark this answers is not in the document]
Comment (vmud) [reply tests/test_cache_interface.py]: Fixed in commit 48529f8. Added `test_cache_interface.py` to the Test Structure section of CLAUDE.md.
